File: 48_Rotate Image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        row, col = len(matrix), len(matrix[0])
        # 1. diagonal reflection (right downwards)
        for r in range(row):
            for c in range(r + 1):
                matrix[r][c], matrix[c][r] = matrix[c][r], matrix[r][c]
        # 2. mirror image: row reflection
        for r in range(row):
            left, right = 0, col - 1
            while left < right:
                matrix[r][left], matrix[r][right] = matrix[r][right], matrix[r][left]
                left += 1
                right -= 1

File: 48_Rotate Image.java
// 时间复杂度：O(n^2 )，其中 n 是 matrix 的行数和列数
// 空间复杂度：O(1)

class Solution {
    public void rotate(int[][] matrix) {
        int row = matrix.length, col = matrix[0].length;

        // 1. reflect right-downwards diagnally
        for (int r = 1; r < row; r++) {
            for (int c = 0; c < r; c++) {
                int tmp = matrix[r][c];
                matrix[r][c] = matrix[c][r];
                matrix[c][r] = tmp;
            }
        }
        // 2. mirror lines
        for (int r = 0; r < row; r++) {
            int left = 0, right = col - 1;
            while (left < right) {
                int tmp = matrix[r][left];
                matrix[r][left++] = matrix[r][right];
                matrix[r][right--] = tmp;
            }
        }
    }
}